### PR TITLE
Unsafe lint will also check for unsafe functions, traits, and implementations

### LINKED ISSUE #22430: unsafe-blocks lint should forbid `unsafe impl`
This program compiles:

```rust
#![forbid(unsafe_blocks)]

unsafe trait Foo { }

struct Bar;

unsafe impl Foo for Bar { }

fn main() { }
```

We should have a lint that forbids *all* use of the unsafe dialect, including `unsafe impl`.  Probably this entails renaming the unsafe-blocks lint.

### PR DESCRIPTION
**Pull request by ipetkov**
This allows warning or forbidding all uses of unsafe code, whereas
previously only unsafe blocks were caught by the lint.

The lint has been renamed from `unsafe-blocks` to `unsafe-code` to
reflect its new purpose.

This is a minor [breaking-change]

Closes #22430

### CODE REVIEW

File: src/librustc/lint/builtin.rs [hunk 1]
```diff
@@ -1269,27 +1269,40 @@ impl LintPass for UnusedUnsafe {
 }
 
 declare_lint! {
-    UNSAFE_BLOCKS,
+    UNSAFE_CODE,
     Allow,
-    "usage of an `unsafe` block"
+    "usage of an `unsafe` code"
```
Comment (kmcallister) [inline src/librustc/lint/builtin.rs]: nit: drop the article `an`

Comment (ipetkov) [inline src/librustc/lint/builtin.rs]: oops, totally missed that, thanks for the catch!

**Comment by ipetkov** [comment]
@huonw I added tests for (both declaring and implementing) unsafe methods of traits. My previous implementation of the lint didn't catch these, so I refactored the code to catch them as well!

**Comment by huonw** [comment]
@bors r+ be0d

Thanks!

**Comment by Manishearth** [comment]
Needs https://github.com/Manishearth/rust/commit/bf13792b319ea0f85900fba134f4102037369f47 to work.

**Comment by ipetkov** [comment]
Thank you for the fix @Manishearth!

**Pull request merged**